internal/ls/lsutil/userpreferences.go

```diff
@@ -308,414 +249,505 @@ const (
-				p.InlayHints.IncludeInlayFunctionParameterTypeHints = parseEnabledBool(v)
-			case "variableTypes":
-				p.InlayHints.IncludeInlayVariableTypeHints = parseEnabledBool(v)
-				p.InlayHints.IncludeInlayVariableTypeHintsWhenTypeMatchesName = parseSuppress(v, "suppressWhenTypeMatchesName")
-			case "propertyDeclarationTypes":
-				p.InlayHints.IncludeInlayPropertyDeclarationTypeHints = parseEnabledBool(v)
-			case "functionLikeReturnTypes":
-				p.InlayHints.IncludeInlayFunctionLikeReturnTypeHints = parseEnabledBool(v)
-			case "enumMemberValues":
-				p.InlayHints.IncludeInlayEnumMemberValueHints = parseEnabledBool(v)
+				setFieldFromValue(field, value)
 			}
-		} else {
-			// non-vscode case
-			p.set(name, v)
 		}
 	}
-}
 
-func (p *UserPreferences) parseReferencesCodeLens(prefs any) {
-	referencesCodeLens, ok := prefs.(map[string]any)
-	if !ok {
-		return
-	}
-	for name, value := range referencesCodeLens {
-		switch name {
-		case "enabled":
-			p.set("referencesCodeLensEnabled", value)
-		case "showOnAllFunctions":
-			p.set("referencesCodeLensShowOnAllFunctions", value)
+	// Process path-based config (VS Code style)
+	for _, info := range infos {
+		if info.path == "" {
+			continue
+		}
+		val, ok := getNestedValue(config, info.path)
+		if !ok {
+			continue
 		}
-	}
-}
 
-func (p *UserPreferences) parseImplementationsCodeLens(prefs any) {
-	implementationsCodeLens, ok := prefs.(map[string]any)
-	if !ok {
-		return
-	}
-	for name, value := range implementationsCodeLens {
-		switch name {
-		case "enabled":
-			p.set("implementationsCodeLensEnabled", value)
-		case "showOnInterfaceMethods":
-			p.set("implementationsCodeLensShowOnInterfaceMethods", value)
-		case "showOnAllClassMethods":
-			p.set("implementationsCodeLensShowOnAllClassMethods", value)
+		field := getFieldByPath(v, info.fieldPath)
+		if info.invert {
+			if b, ok := val.(bool); ok {
+				val = !b
+			}
 		}
+		setFieldFromValue(field, val)
 	}
-}
 
-func (p *UserPreferences) parseSuggest(prefs any) {
-	completionsPreferences, ok := prefs.(map[string]any)
-	if !ok {
-		return
-	}
-	for name, value := range completionsPreferences {
+	// Process direct field names at root level (non-VS Code clients, fourslash tests)
+	for name, value := range config {
+		if name == "unstable" {
+			continue
+		}
+		// Skip known VS Code config sections that are handled by path-based parsing
 		switch name {
-		case "autoImports":
-			p.set("includeCompletionsForModuleExports", value)
-		case "objectLiteralMethodSnippets":
-			if v, ok := value.(map[string]any); ok {
-				p.set("includeCompletionsWithObjectLiteralMethodSnippets", parseEnabledBool(v))
-			}
-		case "classMemberSnippets":
-			if v, ok := value.(map[string]any); ok {
-				p.set("includeCompletionsWithClassMemberSnippets", parseEnabledBool(v))
+		case "preferences", "suggest", "inlayHints", "referencesCodeLens",
+			"implementationsCodeLens", "workspaceSymbols", "format", "tsserver", "tsc", "experimental":
+			continue
```

Comment:
What's an example of a path that doesn't need to be skipped?

Reply:
Yeah, I think this is actually dead code.

Reply:
Ah, so this was here to roundtrip unstable settings, but it isn't doing that right.

I'm going to rethink unstable settings. Probably just make them explicit.

Reply:
Okay, so the sort of problem is that we used to allow user prefs from like, anywhere, so you could have a object with say, `{ "quotePreference": ... }` and that would be read in, which is a valid `UserPreferences` object. But, now over LSP, we are expecting things over `typescript`. So it's maybe a question of what non-vscode people do.